build/extracted-examples/guides/hack/12-generics/01-introduction/Stack.inc.hack

```diff
@@ -4,25 +4,23 @@
 namespace HHVM\UserDocumentation\Guides\Hack\Generics\Introduction\Stack;
 
 class StackUnderflowException extends \Exception {}
+use namespace HH\Lib\C;
 
 class Stack<T> {
   private vec<T> $stack;
-  private int $stackPtr;
 
   public function __construct() {
-    $this->stackPtr = 0;
     $this->stack = vec[];
   }
 
   public function push(T $value): void {
     $this->stack[] = $value;
-    $this->stackPtr++;
   }
 
   public function pop(): T {
-    if ($this->stackPtr > 0) {
-      $this->stackPtr--;
-      return $this->stack[$this->stackPtr];
+    $stack = $this->stack;
+    if (!C\is_empty($stack)) {
+      return C\pop_backx(inout $stack);
```

Comment:
This method does not remove the last element of the stack property. You have made an implicit copy when assigning to a local variable. `C\pop_backx()` is altering this local copy, not the vec in the property.

Reply:
Ah, you're right. I'm just returning the modified local. The reason I made a local was because `C\pop_backx()` said it had to operate on a local property, one marked as `inout`. Passing in the global was making the type checker angry.

Reply:
I could just do something like setting the global from the local, then return the value.

Reply:
@lexidor what do you think of:

```
  private vec<T> $stack;

  public function __construct() {
    $this->stack = vec[];
  }

  public function push(T $value): void {
    $this->stack[] = $value;
  }

  public function pop(): T {
    $stack = $this->stack;
    if (!C\is_empty($stack)) {
      $value = C\pop_backx(inout $stack);
      $this->stack = $stack;
      return $value;
    } else {
      throw new StackUnderflowException();
    }
  }
}
```

Reply:
Yes, that is correct. The typechecker disallows you from passing using a property as the "source" (not the correct term) for an inout. You can reassign the local variable to the property _or_ use a Vector of a vec. I am torn on which decision is best here. On the one hand, the docs clearly favor Hack arrays over collections everywhere else. On the other hand, making a copy to pop and then altering it incurs an `O(N)` cost, since hhvm needs to copy all the elements and increment their refcounts when you write to the local. A Vector is a reference type, so it does not do CoW.

Reply:
Sorry for leading you up a wrong path when I was suggesting you wouldn't need a stackPtr anymore. Doing this with a vec and no stackPtr incurs an O(N) cost on pop.

I have three implementations of a stack, a Vector based, dict based, and vec based stack.

The Vector based is the "best" stack. It does not hold on to elements that are already popped. The garbage collector will be able to collect elements that were previously only referenced in an "unreachable" section of the stack. The dict based stack is a minor correction on the old Stack from the example. By using a dict instead of a vec, push() will always be `[$next_index]=` instead of `[]=`. It is correct, but the memory layout of the elements is not contiguous. The vec based stack is the best you can do with Hack arrays. You need to keep track of the next unused index, since `[]=` must be used when appending to the end of a vec. You can't simply use `[$next_index]=`, this throws an OutOfBoundsException. So you need to pick the right form of the write operation depending on where the next element goes.

I have copied the implementations below.

```HACK
interface StackLike<T> {
    public function push(T $element): void;
    public function pop(): T;
}

final class VectorStack<T> implements StackLike<T> {
    private Vector<T> $elements = Vector {};

    public function push(T $element): void {
        $this->elements->add($element);
    }

    public function pop(): T {
        try {
            return $this->elements->pop();
        } catch (InvalidOperationException $_) {
            throw new StackUnderflowException();
        }
    }
}

final class DictStack<T> implements StackLike<T> {
    private dict<int, T> $elements = dict[];
    private int $stackPtr = 0;

    public function push(T $element): void {
        $this->elements[$this->stackPtr] = $element;
        $this->stackPtr++;
    }

    public function pop(): T {
        if ($this->stackPtr > 0) {
            $this->stackPtr--;
            return $this->elements[$this->stackPtr];
        }
        throw new StackUnderflowException();
    }
}

final class VecStack<T> implements StackLike<T> {
    private vec<T> $elements = vec[];
    private int $size = 0, $capacity = 0;

    public function push(T $element): void {
        if ($this->size === $this->capacity) {
            $this->elements[] = $element;
            $this->capacity++;
        } else {
            $this->elements[$this->size] = $element;
        }
        $this->size++;
    }

    public function pop(): T {
        if ($this->size > 0) {
            $this->size--;
            return $this->elements[$this->size];
        }
        throw new StackUnderflowException();
    }
}

final class StackUnderflowException extends Exception {}

<<__EntryPoint>>
function test(): void {
    $ops = vec<(function(StackLike<int>): mixed)>[
        $s ==> $s->pop(),
        $s ==> $s->push(1),
        $s ==> $s->pop(),
        $s ==> $s->push(1),
        $s ==> $s->push(2),
        $s ==> $s->push(3),
        $s ==> $s->pop(),
        $s ==> $s->pop(),
        $s ==> $s->pop(),
        $s ==> $s->push(1),
        $s ==> $s->push(2),
        $s ==> $s->pop(),
        $s ==> $s->push(3),
        $s ==> $s->pop(),
        $s ==> $s->pop(),
        $s ==> $s->pop(),
    ];

    $s = new VectorStack();
    foreach ($ops as $op) {
        try {
            $res = $op($s);
        } catch (StackUnderflowException $e) {
            $res = \get_class($e);
        }

        if ($res is arraykey) {
            echo $res."\n";
        } else {
            echo "[push]\n";
        }
    }
}
```

Reply:
I feel I need to apologize once again for suggesting using `C\pop_backx()`. When I suggested this on #1171 I was thinking about `return C\pop_backx(inout $this->stack)`. This is not allowed. I misremembered the rules about which arguments qualify. Please do not use `C\pop_backx()` in this Stack example. The dance with the local variable incurs CoW, and therefore makes `->pop()` an order N function. Someone somewhere is going to copy this Stack into their code and blow through their response time budgets with stacks larger than 10,000 elements. I have ran a benchmark of VecStack, VectorStack, and MyStack (called PopBackStack). I timed how long it took to pop a stack of `n` elements until it is empty. The 2^10 rows operate on 1024 elements, the 2^16 rows operate on 65535 elements. The exact numbers are meaningless, but the trend and the comparative results do count.

Running in CLI mode

```
VectorStack   2^10 small 11.14503 ms >> __hhvm_launder_value(523776)
VecStack      2^10 small 6.334611 ms >> __hhvm_launder_value(523776)
PopBackStack  2^10 small 14.475042 ms >> __hhvm_launder_value(523776)
VectorStack   2^11 small 4.915844 ms >> __hhvm_launder_value(2096128)
VecStack      2^11 small 4.439112 ms >> __hhvm_launder_value(2096128)
PopBackStack  2^11 small 20.500871 ms >> __hhvm_launder_value(2096128)
VectorStack   2^12 small 0.314585 ms >> __hhvm_launder_value(8386560)
VecStack      2^12 small 0.289365 ms >> __hhvm_launder_value(8386560)
PopBackStack  2^12 small 38.393394 ms >> __hhvm_launder_value(8386560)
VectorStack   2^13 small 0.569701 ms >> __hhvm_launder_value(33550336)
VecStack      2^13 small 0.63312 ms >> __hhvm_launder_value(33550336)
PopBackStack  2^13 small 143.488665 ms >> __hhvm_launder_value(33550336)
VectorStack   2^14 small 1.611967 ms >> __hhvm_launder_value(134209536)
VecStack      2^14 small 1.313712 ms >> __hhvm_launder_value(134209536)
PopBackStack  2^14 small 545.340548 ms >> __hhvm_launder_value(134209536)
VectorStack   2^15 small 2.074609 ms >> __hhvm_launder_value(536854528)
VecStack      2^15 small 2.45884 ms >> __hhvm_launder_value(536854528)
PopBackStack  2^15 small 2147.56281 ms >> __hhvm_launder_value(536854528)
VectorStack   2^16 small 4.20304 ms >> __hhvm_launder_value(2147450880)
VecStack      2^16 small 4.763756 ms >> __hhvm_launder_value(2147450880)
PopBackStack  2^16 small 7962.39552 ms >> __hhvm_launder_value(2147450880)
```

Running in a webserver with a warm jit (did not speed up PopBackStack)

```
VectorStack   2^10 small 0.09595 ms >> __hhvm_launder_value(523776)
VecStack      2^10 small 0.067733 ms >> __hhvm_launder_value(523776)
PopBackStack  2^10 small 1.962322 ms >> __hhvm_launder_value(523776)
VectorStack   2^11 small 0.110098 ms >> __hhvm_launder_value(2096128)
VecStack      2^11 small 0.121941 ms >> __hhvm_launder_value(2096128)
PopBackStack  2^11 small 7.672828 ms >> __hhvm_launder_value(2096128)
VectorStack   2^12 small 0.185667 ms >> __hhvm_launder_value(8386560)
VecStack      2^12 small 0.221244 ms >> __hhvm_launder_value(8386560)
PopBackStack  2^12 small 33.073805 ms >> __hhvm_launder_value(8386560)
VectorStack   2^13 small 0.360809 ms >> __hhvm_launder_value(33550336)
VecStack      2^13 small 0.630479 ms >> __hhvm_launder_value(33550336)
PopBackStack  2^13 small 125.909002 ms >> __hhvm_launder_value(33550336)
VectorStack   2^14 small 0.700814 ms >> __hhvm_launder_value(134209536)
VecStack      2^14 small 0.847374 ms >> __hhvm_launder_value(134209536)
PopBackStack  2^14 small 487.473917 ms >> __hhvm_launder_value(134209536)
VectorStack   2^15 small 1.515368 ms >> __hhvm_launder_value(536854528)
VecStack      2^15 small 1.660033 ms >> __hhvm_launder_value(536854528)
PopBackStack  2^15 small 2020.03276 ms >> __hhvm_launder_value(536854528)
VectorStack   2^16 small 3.364754 ms >> __hhvm_launder_value(2147450880)
VecStack      2^16 small 3.381403 ms >> __hhvm_launder_value(2147450880)
PopBackStack  2^16 small 8155.71524 ms >> __hhvm_launder_value(2147450880)
```

The bench script

```HACK
<<__EntryPoint>>
function test(): void {
    \header('Content-Type: text/plain');

    for ($i = 3; $i < 7; ++$i) {
        bench(fill(new VectorStack(), $i), $i, 'small');
        bench(fill(new VecStack(), $i), $i, 'small');
        bench(fill(new PopBackStack(), $i), $i, 'small');
    }
}

final class TestObject {
    public function __construct(private int $value) {}

    public function getValue(): int {
        return $this->value;
    }
}

function fill(
    StackLike<TestObject> $stack,
    int $powers_of_ten,
): StackLike<TestObject> {
    $depth = (int)(10 ** $powers_of_ten);

    for ($i = 0; $i < $depth; ++$i) {
        $stack->push(new TestObject($i));
    }

    return $stack;
}

function bench(
    StackLike<TestObject> $stack,
    int $powers_of_ten,
    string $kind_of_object,
): void {
    $start = clock_gettime_ns(CLOCK_MONOTONIC);
    $total = 0;
    try {
        for (; ; ) {
            $total += $stack->pop()->getValue();
        }
    } catch (StackUnderflowException $e) {
        $end = clock_gettime_ns(CLOCK_MONOTONIC);
        echo Str\format(
            "%-13s 10e%d %s %.9g ms >> __hhvm_launder_value(%d)\n",
            \get_class($stack),
            $powers_of_ten,
            $kind_of_object,
            ($end - $start) / 1000000.,
            $total,
        );
    }
}
```

Reply:
For the Vector code, also a good candidate to add to API-Examples for this page, to render an Examples section with that code.

https://docs.hhvm.com/hack/reference/class/HH.Vector/

And the dict and vec implementations might be great to add here, at the bottom of the page: https://docs.hhvm.com/hack/arrays-and-collections/vec-keyset-and-dict

Reply:
@lexidor That benchmark is fascinating. I almost want to formalize and add it somewhere as an example of better understanding why you should are about O(n), but that's probably outside the scope of these docs.

Unless anyone objects, I'll use the Vector code, but thank you very much for your implementation @Wilfred.

Reply:
`Vector`, `Map`, and `Set` are strongly discouraged - we should try really hard to avoid them except when documenting the specific class.

This is both out of a general 'they're replaced by Hack arrays' thing, but also that 'hack has too many array types, when would I use which?' is a common question - the answer is use hack arrays unless you're working on existing code that uses something else.

Reply:
Good point. I'll use the vec implementation.